data_pipeline_api/registry/downloader.py

```diff
@@ -0,0 +1,372 @@
+import itertools
+import logging
+import os
+import urllib
+from collections import defaultdict
+from pathlib import Path
+
+from typing import Dict, Optional, List, Tuple, Any, Union, IO
+
+import yaml
+from fsspec.implementations.sftp import SFTPFileSystem
+
+from data_pipeline_api.registry.common import (
+    get_data,
+    DataRegistryField,
+    DataRegistryTarget,
+    DATA_REGISTRY_URL,
+    DEFAULT_DATA_REGISTRY_URL,
+    DATA_REGISTRY_ACCESS_TOKEN,
+    sort_by_semver,
+    get_on_end_point,
+    get_remote_filesystem_and_path,
+    unique_dicts,
+)
+from data_pipeline_api.metadata import MetadataKey
+
+logger = logging.getLogger(__name__)
+
+
+class Downloader:
+    """
+    Class to handle downloading data products and external objects from the data registry to disk
+    """
+
+    def __init__(
+        self, data_directory: Union[Path, str], data_registry_url: Optional[str] = None, token: Optional[str] = None
+    ) -> None:
+        """
+        :param data_directory: The directory to download data to
+        :param data_registry_url: base url of the data registry
+        :param token: personal access token
+        """
+        self._data_directory = Path(data_directory)
+        self._data_registry_url: str = data_registry_url or os.environ.get(DATA_REGISTRY_URL, DEFAULT_DATA_REGISTRY_URL)
+        self._token: str = token or os.environ.get(DATA_REGISTRY_ACCESS_TOKEN)
+
+        self._data_products: List[Dict[Tuple[str, str], str]] = []
+        self._external_objects: List[Dict[Tuple[str, str], str]] = []
+        self._resolved_data_products: List[Dict[Tuple[str, str], Any]] = []
```

Comment:
A triviality, but it might be nice to make a more descriptive type alias for `Dict[Tuple[str, str], Any]`, given how many times it is used.